# IDX Data Analyst Internship 

### Week 1- downloading the files, setting up python, setting up git, cleaning data.

In [5]:
#Imports
import pandas as pd
import crmls_listed
import crmls_sold


Total 31410 records exported to CRMLSListing202602.csv
Total 19106 records exported to CRMLSSold202602.csv


In [6]:
import glob

In [7]:
df_listings= pd.read_csv("CRMLSListing202602.csv")
df_sold = pd.read_csv("CRMLSSold202602.csv")

### Displaying the data frames 

In [8]:
glob.glob("*.csv")

['CRMLSListing202602.csv', 'CRMLSSold202602.csv']

In [9]:
len(glob.glob("CRMLSSold*.csv"))

1

In [10]:
sold_files = glob.glob("CRMLSSold*.csv")
listed_files = glob.glob("CRMLSListing*.csv")

print(len(sold_files))
print(len(listed_files))

1
1


In [11]:
import os
os.listdir()

['.DS_Store',
 'crmls_sold.py',
 'CRMLSListing202602.csv',
 '__pycache__',
 'README.md',
 '.gitignore',
 'CRMLSSold202602.csv',
 '.git',
 'crmls_listed.py',
 'analysis.ipynb',
 'raw']

In [12]:
sold_files = glob.glob("raw/CRMLSSold*.csv")
df_sold = pd.concat([pd.read_csv(f, low_memory= False) for f in sold_files], ignore_index=True)


In [13]:
df_listings = pd.concat(
    [pd.read_csv(f, low_memory=False) for f in listed_files],
    ignore_index=True)

In [14]:
# Looking at the original data set 
df_sold

,BuyerAgentAOR,ListAgentAOR,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,ListAgentEmail,...,PostalCode,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,latfilled,lonfilled,OriginatingSystemName,OriginatingSystemSubName,BuyerAgencyCompensationType,BuyerAgencyCompensation
0,Arcadia,Arcadia,NaN,False,NaN,NaN,False,1130000.0,538338723,espi.bagwell@kw.com,...,91016,0.0,14390.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Conejo,Conejo,NaN,True,NaN,NaN,NaN,1795000.0,535502637,david@mabryflp.com,...,93021,NaN,666032.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,OrangeCounty,OrangeCounty,NaN,True,NaN,NaN,NaN,99000.0,495549184,Jaredengberg@rocketmail.com,...,92283,0.0,871200.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,SouthwestRiversideCounty,SouthwestRiversideCounty,NaN,True,NaN,NaN,NaN,40000.0,446937009,Julie@whitehornhomes.com,...,92530,0.0,16553.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,HighDesert,HighDesert,NaN,True,NaN,NaN,NaN,59900.0,446912267,nadinesells4u@gmail.com,...,92327,0.0,275800.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
532590,NaN,NaN,NaN,NaN,NaN,NaN,NaN,649000.0,1008475397,chiggamasta424@gmail.com,...,90805,NaN,4682.0,NaN,NaN,NaN,NaN,NaN,Item1,2.5
532591,NaN,NaN,NaN,True,NaN,NaN,NaN,75689.0,1003434877,jhisquierdo@lee-associates.com,...,92371,NaN,946123.0,NaN,NaN,NaN,NaN,NaN,Item1,4.0
532592,NaN,NaN,NaN,True,NaN,NaN,NaN,275000.0,1002713752,MLS@DaltonRealtors.com,...,92082,0.0,619859.0,NaN,NaN,NaN,NaN,NaN,Item1,2.0
532593,NaN,NaN,NaN,False,NaN,NaN,False,10000.0,1002447928,Tazzie@CoronadoPremierProperties.com,...,92118,0.0,NaN,NaN,NaN,NaN,NaN,NaN,Item,100.0


In [15]:
#Looking at the original data set
df_listings

,OriginalListPrice,ListingKey,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,...,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,BuyerOfficeName.1,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,UnparsedAddress.1
0,1100000.0,1157171473,NaN,NaN,Cassandra,Diep,34.076588,-118.085541,3821 Earle Avenue,Residential,...,1.0,False,2.0,Rosemead,91770,NaN,0.0,7212.0,NaN,3821 Earle Avenue
1,2130000.0,1157170900,2026-04-09,2130000.0,Briant,Patterson,33.651866,-117.898412,2342 Westminster,Residential,...,6.0,False,2.0,Newport Mesa Unified,92627,Harcourts Prime Properties,0.0,8276.0,NaN,2342 Westminster
2,1199000.0,1157170763,NaN,NaN,Audrey,Huang,34.095744,-118.137994,312 N Marguerita Avenue,Residential,...,NaN,NaN,2.0,NaN,91801,NaN,NaN,7527.0,NaN,312 N Marguerita Avenue
3,530000.0,1157170224,2026-04-07,530000.0,Nick,Rogers,33.036771,-116.868485,708 12th street,Residential,...,NaN,False,2.0,NaN,92065,Coldwell Banker Realty,NaN,NaN,NaN,708 12th street
4,225000.0,1157169737,NaN,NaN,Tina,Quiroz,33.861935,-116.254086,0 Vacant Land,Land,...,NaN,False,NaN,NaN,92241,NaN,0.0,210395.0,NaN,0 Vacant Land
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
31405,4950.0,1149458718,NaN,NaN,Alphonso,Lascano,34.044402,-118.365074,1716 S Genesee Avenue,ResidentialLease,...,NaN,False,NaN,NaN,90019,NaN,NaN,4886.0,NaN,1716 S Genesee Avenue
31406,1699.0,1146439434,NaN,NaN,Gene,Smith,38.115612,-122.267146,124 Benson Ave,ResidentialLease,...,NaN,False,1.0,NaN,94590,NaN,NaN,15681.0,NaN,124 Benson Ave
31407,350000.0,1146104836,NaN,NaN,Jason,Lipovsky,33.139128,-117.053881,475 N Midway Drive 169,Residential,...,0.0,NaN,0.0,Escondido Union,92027,NaN,400.0,261360.0,NaN,475 N Midway Drive 169
31408,1325.0,1107246684,NaN,NaN,Kathy,Lund,32.572312,-117.034778,4589 Casa nova ct,ResidentialLease,...,5.0,False,2.0,Other,92154,NaN,0.0,2224.0,NaN,4589 Casa nova ct


Deliverable – Week 1
Submit a .py script that concatenates all monthly files from January 2024 through the most recently
completed calendar month into two combined datasets (listings and sold), filters both to PropertyType ==
‘Residential’ only, and saves them as new CSVs. Include comments confirming row counts before and
after concatenation and before and after the Residential filter

#### Creating new data frames named df_sold_residential, and df_listings_residential, that contain only observations of residential homes. 

In [41]:
# filtering csv files to property type == Residential.
df_sold_residential = df_sold[df_sold["PropertyType"]=="Residential"]
df_listings_residential = df_listings[df_listings["PropertyType"] == "Residential"]

#### Looking at the number of rows and columns in the data sets. 

In [36]:
df_listings.shape

(31410, 81)

In [37]:
df_sold.shape

(532595, 84)

In [43]:
df_listings_residential.shape

(19997, 81)

In [44]:
df_sold_residential.shape

(359471, 84)

##### Looking at the columns and the number of rows of the new and old data set. The original data sets are significantly larger compared to the new filtered data sets. 
##### df_listings has 31410 rows and 81 colums, df_sold has 532595 rows and 84 columns. 
##### I expect the new data frames to have the same amount of columns which it did, and the rows to reduce, 
##### df_listings_residential has 19997 rows and 81 columns, and df_sold_residential has 359471 rows and 84 columns. 

#### Deliverable – Weeks 2–3
Submit a .py script documenting unique property types found, the filtering logic applied, and a null-count
summary table. Include a missing value report flagging any columns above 90% null. Produce a numeric
distribution summary (min, max, mean, median, percentiles) for ClosePrice, LivingArea, and
DaysOnMarket. Save the filtered dataset as a new CSV.

In [16]:
#Displaying column names to filter data and create new data frames.
df_sold.columns

Index(['BuyerAgentAOR', 'ListAgentAOR', 'Flooring', 'ViewYN', 'WaterfrontYN',
       'BasementYN', 'PoolPrivateYN', 'OriginalListPrice', 'ListingKey',
       'ListAgentEmail', 'CloseDate', 'ClosePrice', 'ListAgentFirstName',
       'ListAgentLastName', 'Latitude', 'Longitude', 'UnparsedAddress',
       'PropertyType', 'LivingArea', 'ListPrice', 'DaysOnMarket',
       'ListOfficeName', 'BuyerOfficeName', 'CoListOfficeName',
       'ListAgentFullName', 'CoListAgentFirstName', 'CoListAgentLastName',
       'BuyerAgentMlsId', 'BuyerAgentFirstName', 'BuyerAgentLastName',
       'FireplacesTotal', 'AssociationFeeFrequency', 'AboveGradeFinishedArea',
       'ListingKeyNumeric', 'MLSAreaMajor', 'TaxAnnualAmount',
       'CountyOrParish', 'MlsStatus', 'ElementarySchool', 'AttachedGarageYN',
       'ParkingTotal', 'BuilderName', 'PropertySubType', 'LotSizeAcres',
       'SubdivisionName', 'BuyerOfficeAOR', 'YearBuilt', 'StreetNumberNumeric',
       'ListingId', 'BathroomsTotalInteger', 'City', '

In [17]:
df_listings.columns

Index(['OriginalListPrice', 'ListingKey', 'CloseDate', 'ClosePrice',
       'ListAgentFirstName', 'ListAgentLastName', 'Latitude', 'Longitude',
       'UnparsedAddress', 'PropertyType', 'LivingArea', 'ListPrice',
       'DaysOnMarket', 'ListOfficeName', 'BuyerOfficeName', 'CoListOfficeName',
       'ListAgentFullName', 'CoListAgentFirstName', 'CoListAgentLastName',
       'BuyerAgentMlsId', 'BuyerAgentFirstName', 'BuyerAgentLastName',
       'FireplacesTotal', 'AssociationFeeFrequency', 'AboveGradeFinishedArea',
       'ListingKeyNumeric', 'MLSAreaMajor', 'TaxAnnualAmount',
       'CountyOrParish', 'PropertyType.1', 'MlsStatus', 'ElementarySchool',
       'ListAgentFirstName.1', 'AttachedGarageYN', 'ParkingTotal',
       'BuilderName', 'PropertySubType', 'LotSizeAcres', 'SubdivisionName',
       'BuyerOfficeAOR', 'YearBuilt', 'DaysOnMarket.1', 'StreetNumberNumeric',
       'LivingArea.1', 'ListingId', 'BathroomsTotalInteger', 'City', 'TaxYear',
       'BuildingAreaTotal', 'BedroomsTota

### Creatign new data frames with our variables of interest:

#### Key MLS Fields

List Price / Close Price — Transaction pricing data

Square Footage — Property size metrics

(LivingArea)
Bedrooms / Bathrooms — Property configuration

Property Type — Residential, Condo, etc.

Listing Agent / Brokerage — Agent and firm attribution

Days on Market (DaysOnMarket) — Time-to-sell metric

In [18]:
df_sold_new = df_sold[["ClosePrice","ListPrice", "LotSizeSquareFeet","BedroomsTotal", "BathroomsTotalInteger","PropertyType", "BuyerAgentAOR", "ListAgentAOR", "DaysOnMarket"]].copy()
df_sold_new

,ClosePrice,ListPrice,LotSizeSquareFeet,BedroomsTotal,BathroomsTotalInteger,PropertyType,BuyerAgentAOR,ListAgentAOR,DaysOnMarket
0,1090000.0,1130000.0,14390.0,3.0,4.0,Residential,Arcadia,Arcadia,26
1,1000000.0,1200000.0,666032.0,NaN,0.0,Land,Conejo,Conejo,1134
2,12999.0,14999.0,871200.0,NaN,NaN,Land,OrangeCounty,OrangeCounty,1108
3,40000.0,40000.0,16553.0,NaN,NaN,Land,SouthwestRiversideCounty,SouthwestRiversideCounty,1517
4,69000.0,79900.0,275800.0,NaN,NaN,Land,HighDesert,HighDesert,1476
...,...,...,...,...,...,...,...,...,...
532590,450000.0,549000.0,4682.0,NaN,NaN,CommercialSale,NaN,NaN,769
532591,75689.0,75689.0,946123.0,NaN,NaN,Land,NaN,NaN,547
532592,100000.0,99000.0,619859.0,NaN,NaN,Land,NaN,NaN,846
532593,8300.0,8300.0,NaN,3.0,4.0,ResidentialLease,NaN,NaN,58


In [19]:
df_listing_new= df_listings[["ListPrice","ClosePrice", "LotSizeSquareFeet","LivingArea", "BedroomsTotal", "BathroomsTotalInteger","PropertyType", "DaysOnMarket", "ListAgentFirstName","ListAgentLastName", "ListAgentFullName","ListOfficeName","CoListOfficeName"]].copy()
df_listing_new

,ListPrice,ClosePrice,LotSizeSquareFeet,LivingArea,BedroomsTotal,BathroomsTotalInteger,PropertyType,DaysOnMarket,ListAgentFirstName,ListAgentLastName,ListAgentFullName,ListOfficeName,CoListOfficeName
0,1100000.0,NaN,7212.0,2410.0,4.0,3.0,Residential,0,Cassandra,Diep,Cassandra Diep,eXp Realty of California Inc,eXp Realty of California Inc
1,2130000.0,2130000.0,8276.0,1903.0,6.0,4.0,Residential,0,Briant,Patterson,Briant Patterson,Pacific Sotheby's Int'l Realty,NaN
2,1199000.0,NaN,7527.0,1925.0,4.0,2.0,Residential,0,Audrey,Huang,Audrey Huang,COMPASS,NaN
3,530000.0,530000.0,NaN,1020.0,2.0,1.0,Residential,0,Nick,Rogers,Nick Rogers,Coldwell Banker Realty,NaN
4,225000.0,NaN,210395.0,NaN,NaN,NaN,Land,0,Tina,Quiroz,Tina Quiroz,Seven Gables Real Estate,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
31405,4950.0,NaN,4886.0,1008.0,2.0,2.0,ResidentialLease,15,Alphonso,Lascano,Alphonso Lascano,Carolwood Estates,Carolwood Estates
31406,1650.0,NaN,15681.0,7920.0,2.0,1.0,ResidentialLease,49,Gene,Smith,Gene Smith,First Republic Real Estate,NaN
31407,335000.0,NaN,261360.0,1032.0,2.0,1.0,Residential,76,Jason,Lipovsky,Jason Lipovsky,Venture Pacific Properties,NaN
31408,1149.0,NaN,2224.0,2224.0,1.0,1.0,ResidentialLease,359,Kathy,Lund,Kathy Lund,Integrity First Realty & Loans,Integrity First Realty & Loans


In [20]:
df_listing_new.describe()

,ListPrice,ClosePrice,LotSizeSquareFeet,LivingArea,BedroomsTotal,BathroomsTotalInteger,DaysOnMarket
count,3.132000e+04,1.065700e+04,2.873900e+04,27137.000000,27254.000000,28523.000000,31410.000000
mean,1.023248e+06,8.242395e+05,3.634591e+05,1939.234790,3.153115,2.534165,22.314008
std,2.181693e+06,1.121470e+06,1.542362e+07,1628.599255,2.035913,2.031149,19.683160
min,1.000000e+00,1.770000e+00,0.000000e+00,0.000000,0.000000,0.000000,-10.000000
25%,1.348000e+05,5.200000e+03,5.306000e+03,1175.000000,2.000000,2.000000,7.000000
50%,6.488580e+05,6.370000e+05,7.642000e+03,1606.000000,3.000000,2.000000,18.000000
75%,1.178000e+06,1.150000e+06,1.850000e+04,2266.000000,4.000000,3.000000,37.000000
max,9.135000e+07,3.026000e+07,1.897866e+09,94619.000000,149.000000,149.000000,710.000000


In [21]:
df_sold_new.describe()

,ClosePrice,ListPrice,LotSizeSquareFeet,BedroomsTotal,BathroomsTotalInteger,DaysOnMarket
count,5.325900e+05,5.317780e+05,4.830150e+05,497152.000000,507915.000000,532595.00000
mean,8.800521e+05,8.437559e+05,4.523878e+05,3.082136,2.445110,42.33038
std,5.378529e+06,1.275921e+06,1.925662e+07,1.388929,1.413445,69.37806
min,0.000000e+00,0.000000e+00,0.000000e+00,0.000000,0.000000,-288.00000
25%,6.500000e+04,7.500000e+04,5.175000e+03,2.000000,2.000000,8.00000
50%,6.339885e+05,6.350000e+05,7.276000e+03,3.000000,2.000000,21.00000
75%,1.075000e+06,1.070000e+06,1.276800e+04,4.000000,3.000000,52.00000
max,9.895000e+08,1.375000e+08,2.208318e+09,123.000000,175.000000,12430.00000


In [22]:
df_listing_new.head()

,ListPrice,ClosePrice,LotSizeSquareFeet,LivingArea,BedroomsTotal,BathroomsTotalInteger,PropertyType,DaysOnMarket,ListAgentFirstName,ListAgentLastName,ListAgentFullName,ListOfficeName,CoListOfficeName
0,1100000.0,NaN,7212.0,2410.0,4.0,3.0,Residential,0,Cassandra,Diep,Cassandra Diep,eXp Realty of California Inc,eXp Realty of California Inc
1,2130000.0,2130000.0,8276.0,1903.0,6.0,4.0,Residential,0,Briant,Patterson,Briant Patterson,Pacific Sotheby's Int'l Realty,NaN
2,1199000.0,NaN,7527.0,1925.0,4.0,2.0,Residential,0,Audrey,Huang,Audrey Huang,COMPASS,NaN
3,530000.0,530000.0,NaN,1020.0,2.0,1.0,Residential,0,Nick,Rogers,Nick Rogers,Coldwell Banker Realty,NaN
4,225000.0,NaN,210395.0,NaN,NaN,NaN,Land,0,Tina,Quiroz,Tina Quiroz,Seven Gables Real Estate,NaN


In [23]:
df_sold_new.head()

,ClosePrice,ListPrice,LotSizeSquareFeet,BedroomsTotal,BathroomsTotalInteger,PropertyType,BuyerAgentAOR,ListAgentAOR,DaysOnMarket
0,1090000.0,1130000.0,14390.0,3.0,4.0,Residential,Arcadia,Arcadia,26
1,1000000.0,1200000.0,666032.0,NaN,0.0,Land,Conejo,Conejo,1134
2,12999.0,14999.0,871200.0,NaN,NaN,Land,OrangeCounty,OrangeCounty,1108
3,40000.0,40000.0,16553.0,NaN,NaN,Land,SouthwestRiversideCounty,SouthwestRiversideCounty,1517
4,69000.0,79900.0,275800.0,NaN,NaN,Land,HighDesert,HighDesert,1476


### Cleaning the data 
##### 1) Looking at the missing values, which to remove and which to keep.
##### 2) changing any quatitative variable with letters inside. 

In [24]:
df_sold_new.isna().sum()


ClosePrice                   5
ListPrice                  817
LotSizeSquareFeet        49580
BedroomsTotal            35443
BathroomsTotalInteger    24680
PropertyType                 0
BuyerAgentAOR            68027
ListAgentAOR             68027
DaysOnMarket                 0
dtype: int64

In [25]:
df_sold_new[df_sold_new["BathroomsTotalInteger"] == 0]

,ClosePrice,ListPrice,LotSizeSquareFeet,BedroomsTotal,BathroomsTotalInteger,PropertyType,BuyerAgentAOR,ListAgentAOR,DaysOnMarket
1,1000000.0,1200000.0,666032.0,NaN,0.0,Land,Conejo,Conejo,1134
23,945000.0,995000.0,9365.0,NaN,0.0,ResidentialIncome,Mlslistings,Mlslistings,0
28,500000.0,500000.0,436.0,NaN,0.0,CommercialSale,Mlslistings,Mlslistings,0
29,27000.0,27000.0,8276.0,NaN,0.0,Land,CaliforniaDesert,CaliforniaDesert,0
68,17500000.0,17500000.0,9626760.0,NaN,0.0,CommercialSale,Mlslistings,Mlslistings,0
...,...,...,...,...,...,...,...,...,...
532514,26000.0,25812.0,28000.0,NaN,0.0,CommercialLease,NaN,NaN,525
532558,38000.0,55000.0,18020.0,NaN,0.0,Land,NaN,NaN,630
532570,280000.0,298000.0,847678.0,NaN,0.0,Land,NaN,NaN,710
532583,195000.0,199000.0,24393.6,NaN,0.0,Land,NaN,NaN,747


In [26]:
df_sold_new[(df_sold_new["PropertyType"] == "Land") | (df_sold_new["BedroomsTotal"] == "NaN")]

,ClosePrice,ListPrice,LotSizeSquareFeet,BedroomsTotal,BathroomsTotalInteger,PropertyType,BuyerAgentAOR,ListAgentAOR,DaysOnMarket
1,1000000.0,1200000.0,666032.0,NaN,0.0,Land,Conejo,Conejo,1134
2,12999.0,14999.0,871200.0,NaN,NaN,Land,OrangeCounty,OrangeCounty,1108
3,40000.0,40000.0,16553.0,NaN,NaN,Land,SouthwestRiversideCounty,SouthwestRiversideCounty,1517
4,69000.0,79900.0,275800.0,NaN,NaN,Land,HighDesert,HighDesert,1476
5,15000.0,15599.0,108900.0,NaN,NaN,Land,Southland,Southland,2414
...,...,...,...,...,...,...,...,...,...
532587,215000.0,229000.0,152024.0,NaN,NaN,Land,NaN,NaN,761
532588,30000.0,59000.0,1293732.0,NaN,NaN,Land,NaN,NaN,151
532591,75689.0,75689.0,946123.0,NaN,NaN,Land,NaN,NaN,547
532592,100000.0,99000.0,619859.0,NaN,NaN,Land,NaN,NaN,846


#### looking at the data, most of the missing values for bedrooms are because of the property being land and not a building. Therefore it is not something that I would remove from our data set. 

In [27]:
df_listing_new.isna().sum()

ListPrice                   90
ClosePrice               20753
LotSizeSquareFeet         2671
LivingArea                4273
BedroomsTotal             4156
BathroomsTotalInteger     2887
PropertyType                 0
DaysOnMarket                 0
ListAgentFirstName         154
ListAgentLastName            1
ListAgentFullName           17
ListOfficeName               0
CoListOfficeName         25034
dtype: int64

#### Identify number of rows and columns
#### Review column data types
#### Identify high-missing columns
#### Separate market analysis fields from metadata fields

In [ ]:
#identify number of rows and columns
df_listings_new.shape

(31410, 81)

In [31]:
#Identify number of rows and columns
df_sold.shape

(532595, 84)

git add .
git commit -m "your message"
git pull origin main --rebase
git push origin main